File: FirefighterStats/Server/Entities/FirefighterActivities/Intervention.cs
// -----------------------------------------------------------------------
//  <copyright project="FirefighterStats.Server" file="Intervention.cs" company="syuko">
//  Copyright (c) syuko. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------

namespace FirefighterStats.Server.Entities.FirefighterActivities;

using System.ComponentModel.DataAnnotations;
using FirefighterStats.Server.Helpers.Calculators;
using FirefighterStats.Shared.IndemnitySlip.FirefighterActivities;
using FirefighterStats.Shared.Utils;
using JetBrains.Annotations;

[UsedImplicitly]
public class Intervention
{
    private static readonly Percentage s_dayRate = "100%";

    private static readonly Percentage s_nightRate = "200%";

    private static readonly Percentage s_specialRate = "150%";

    private Calculator _calculator = new CalculatorV2();

    private ECalculatorVersion _calculatorVersion;

    private DateTime _endDateTime = DateTime.MinValue;

    private DateTime _startDateTime = DateTime.MinValue;

    public double Amount => Math.Round((DayHours * s_dayRate + NightHours * s_nightRate + SpecialHours * s_specialRate) * UnitAmount, 2);

    public ECalculatorVersion CalculatorVersion
    {
        get => _calculatorVersion;

        set
        {
            _calculatorVersion = value;

            _calculator = Calculator.CreateCalculator(value);
        }
    }

    public double DayHours { get; private set; }

    public required DateTime EndDateTime
    {
        get => _endDateTime;

        set
        {
            _endDateTime = value;
            CalculateHours();
        }
    }

    [Key]
    public required string Id { get; set; }

#pragma warning disable CS8618
    public IndemnitySlip IndemnitySlip { get; set; }
#pragma warning restore CS8618

    public required EInterventionType InterventionType { get; set; }

    public double NightHours { get; private set; }

    public required int Number { get; set; }

    public double SpecialHours { get; private set; }

    public required DateTime StartDateTime
    {
        get => _startDateTime;

        set
        {
            _startDateTime = value;
            CalculateHours();
        }
    }

    public required string Title { get; set; }

    public double TotalHours => DayHours + NightHours + SpecialHours;

    public required double UnitAmount { get; set; }

    private void CalculateHours()
    {
        (DayHours, SpecialHours, NightHours) = _calculator.CalculateHours(StartDateTime, EndDateTime);
    }
}

File: FirefighterStats/Client/src/types/Intervention.ts
import ECalculatorVersion from './ECalculatorVersion';
import EInterventionType from './EInterventionType';
import { DefaultInterventionCalculator } from './InterventionCalculator';

type Intervention = {
    amount: number;
    dayHours: number;
    endDateTime: Date;
    id: string;
    interventionType: keyof typeof EInterventionType;
    calculatorVersion: keyof typeof ECalculatorVersion;
    nightHours: number;
    number: number;
    specialHours: number;
    startDateTime: Date;
    title: string;
    totalHours: number;
    unitAmount: number;
};

export type NewIntervention = Omit<Intervention, 'amount' | 'dayHours' | 'id' | 'nightHours' | 'specialHours' | 'totalHours'>;

export function isIntervention(data: any): data is Intervention {
    return (
        typeof data.amount === 'number' &&
        typeof data.dayHours === 'number' &&
        typeof data.endDateTime === 'string' &&
        typeof data.id === 'string' &&
        Object.keys(EInterventionType).includes(data.interventionType) &&
        Object.keys(ECalculatorVersion).includes(data.calculatorVersion) &&
        typeof data.nightHours === 'number' &&
        typeof data.number === 'number' &&
        typeof data.specialHours === 'number' &&
        typeof data.startDateTime === 'string' &&
        typeof data.title === 'string' &&
        typeof data.totalHours === 'number' &&
        typeof data.unitAmount === 'number'
    );
}

export function getNewInterventionPreview(newIntervention: NewIntervention): Intervention {
    const interventionCalculator = DefaultInterventionCalculator;

    const hours = interventionCalculator.calculateHours(newIntervention.startDateTime, newIntervention.endDateTime);
    const amount = interventionCalculator.calculateAmount(newIntervention.unitAmount, hours);

    return {
        ...newIntervention,
        amount,
        dayHours: hours.dayTime,
        nightHours: hours.nightTime,
        specialHours: hours.specialTime,
        totalHours: hours.dayTime + hours.nightTime + hours.specialTime,
        id: `new-intervention-${newIntervention.number}-preview`,
    };
}

export default Intervention;
